File: python/euler/bin/p0076.py
# project euler: problem 76

#   another version of problem 31
#
#   coins: 1, 2, 3, ..., 99
#   total: 100

from collections.abc import Iterable


def compute(coins: Iterable[int], target: int) -> str:
    tbl = [0] * (target + 1)
    tbl[0] = 1

    for c in coins:
        for i in range(c, target + 1):
            tbl[i] += tbl[i - c]

    return str(tbl[target])


def solve() -> str:
    return compute(list(range(1, 100)), 100)

File: typescript/euler/bin/p0076.ts
// project euler: problem 76

/*
  another version of problem 31

  coins: 1, 2, 3, ..., 99
  total: 100
*/

import { range } from "../lib/util.ts";

export const compute = (coins: number[], target: number): string => {
  const tbl = new Array(target + 1);
  tbl.fill(0);
  tbl[0] = 1;

  for (const c of coins) {
    for (const i of range(c, target + 1)) {
      tbl[i] += tbl[i - c];
    }
  }

  return String(tbl[target]);
};

export const solve = (): string => compute(range(1, 100), 100);
